### PR TITLE
fix unnecessary warning when using bitand with boolean operators

### LINKED ISSUE #77601: Use of bitand on boolean operands
The following code gives a warning as expected:
```cpp
bool test();
int main() {
    return test() & test();
}
```
`warning: use of bitwise '&' with boolean operands [-Wbitwise-instead-of-logical])`

However, this code also gives the same warning:
```cpp
bool test();
int main() {
    return test() bitand test();
}
```

I believe it would be best not to give the warning in the second case. There is no risk of the second case being accidental typo. It would be a nice way of signaling the bitwise AND is intentional to the compiler, rather than casting one of the bools to an int.

### PR DESCRIPTION
**Pull request by 11happy**
**Overview:**
This pull request fixes #77601 where using the `bitand` operator with boolean operands should not trigger the warning, as it would indicate an intentional use of bitwise AND rather than a typo or error. 

**Testing:**
- Tested the updated code.
- Verified that other functionalities remain unaffected.
Screenshot from 2024-02-16 14-25-58

**Dependencies:**
- No dependencies on other pull requests.

**CC:**
- @dtcxzyw

### CODE REVIEW

**Comment by 11happy** [comment]
Humble Ping!

File: clang/lib/Sema/SemaChecking.cpp [hunk 1]
```diff
@@ -16191,12 +16191,24 @@ static void AnalyzeImplicitConversions(
...
-                 BO->getOperatorLoc(),
-                 (BO->getOpcode() == BO_And ? "&&" : "||"));
-      S.Diag(BO->getBeginLoc(), diag::note_cast_operand_to_int);
+      clang::SourceManager &SM = S.getSourceManager();
+      clang::LangOptions LO = S.getLangOpts();
+      clang::SourceLocation BLoc = BO->getOperatorLoc();
+      clang::SourceLocation ELoc =
+          clang::Lexer::getLocForEndOfToken(BLoc, 0, SM, LO);
+      llvm::StringRef SR = clang::Lexer::getSourceText(
+          clang::CharSourceRange::getTokenRange(BLoc, ELoc), SM, LO);
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaChecking.cpp:16200]:
```suggestion
      SourceManager &SM = S.getSourceManager();
      const LangOptions &LO = S.getLangOpts();
      SourceLocation BLoc = BO->getOperatorLoc();
      SourceLocation ELoc =
          Lexer::getLocForEndOfToken(BLoc, 0, SM, LO);
      StringRef SR = Lexer::getSourceText(
          CharSourceRange::getTokenRange(BLoc, ELoc), SM, LO);
```
You shouldn't need to fully qualify any of these from here. Also, no need to copy the `LangOptions` object.

File: clang/lib/Sema/SemaChecking.cpp [hunk 2]
```diff
@@ -16191,12 +16191,24 @@ static void AnalyzeImplicitConversions(
...
+
+      if (SR.str() == "&" || SR.str() == "|") {
+
+        S.Diag(BO->getBeginLoc(), diag::warn_bitwise_instead_of_logical)
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaChecking.cpp:16204]:
```suggestion

      // To reduce false positives, only issue the diagnostic if the operator is explicitly spelled as a
      // punctuator. This suppresses the diagnostic when using 'bitand' or 'bitor' either as keywords
      // in C++ or as macros in C, along with other macro spellings the user might invent.
      if (SR.str() == "&" || SR.str() == "|") {
        S.Diag(BO->getBeginLoc(), diag::warn_bitwise_instead_of_logical)
```
You may need to adjust for 80-col limits.

File: clang/test/Sema/warn-bitwise-and-bool.c [hunk 3]
```diff
@@ -45,8 +45,8 @@ void test(boolean a, boolean b, int *p, volatile int *q, int i) {
   b = bar() & (i > 4);
   b = (i == 7) & foo();
 #ifdef __cplusplus
-  b = foo() bitand bar(); // expected-warning {{use of bitwise '&' with boolean operands}}
-                          // expected-note@-1 {{cast one or both operands to int to silence this warning}}
+  b = foo() bitand bar(); // Ok, no warning expected
+
```
Comment (AaronBallman) [inline clang/test/Sema/warn-bitwise-and-bool.c:49]: You should also add test coverage for C with a macro definition for `bitand` and `bitor`, and it'd be good to add a test (both C and C++) that does something like:
```
#define my_fancy_bit_and &
b = foo() my_fancy_bit_and bar();
```
to show this intentionally works with user-defined macros as well.

**Review by AaronBallman — COMMENTED** [review]
Thank you for the improvement! Please be sure to add a release note about the changes as well.

File: clang/lib/Sema/SemaChecking.cpp (hunk 1 above)
Comment (11happy) [reply clang/lib/Sema/SemaChecking.cpp]: done

File: clang/lib/Sema/SemaChecking.cpp (hunk 2 above)
Comment (11happy) [reply clang/lib/Sema/SemaChecking.cpp]: done

File: clang/test/Sema/warn-bitwise-and-bool.c (hunk 3 above)
Comment (11happy) [reply clang/test/Sema/warn-bitwise-and-bool.c]: done

**Comment by 11happy** [comment]
currently the macro definition for both `&` and `bitand` is not giving any warning however should the `&` one warn? or current working is correct behaviour as expected?

**Comment by AaronBallman** [comment]
> currently the macro definition for both `&` and `bitand` is not giving any warning however should the `&` one warn? or current working is correct behaviour as expected?

The current behavior is what I'd expect. The idea being: if the user is using `&` directly, it's not clear whether they accidentally meant to use `&&` instead, but if they use something with an identifier (a macro, whether it's named `bitand` or not), then there's very little chance they meant `&&` instead.

File: clang/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -250,6 +250,11 @@ Improvements to Clang's diagnostics
   such as attempting to call ``free`` on an unallocated object. Fixes
   `#79443 <https://github.com/llvm/llvm-project/issues/79443>`_.
 
+- Clang now does not warn in cases where bitand operator is
+  intentionally used with boolean operands, distinguishing it
+  from potential typographical errors or unintended bitwise operations.
+  Fixes (`#77601 <https://github.com/llvm/llvm-project/issues/77601>`)
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:256]:
```suggestion
- Clang no longer warns when the ``bitand`` operator is used with boolean
  operands, distinguishing it from potential typographical errors or unintended
  bitwise operations. Fixes #GH77601.
```

**Review by AaronBallman — APPROVED** [review]
LGTM with a small tweak to the release notes.

**Pull request merged**